scripts/trim_indel_ends.py

```diff
@@ -0,0 +1,402 @@
+    """
+    all_edits = extract_edits(cigar_str)
+    prefix_indels = []
+    suffix_indels = []
+    i = 0
+    for i in range(len(all_edits)):
+        if all_edits[i][-1] != "I" and all_edits[i][-1] != "D":
+            break
+        prefix_indels.append(all_edits[i])
+    j = len(all_edits)-1
+    for j in range(len(all_edits)-1, -1, -1):
+        if all_edits[j][-1] != "I" and all_edits[j][-1] != "D":
+            break
+        suffix_indels.append(all_edits[j])
+    for idx in range(i,j+1):
+        if all_edits[idx][-1] != "=" and mode == "strict":
+            return ("", 0), ("",0)
+
+    # return prefix_indels, suffix_indels[::-1]
+    return get_len_type(prefix_indels), get_len_type(suffix_indels[::-1])
+
+def trim_seq(prefix_type, prefix_len, suffix_type, suffix_len, seq, num, threshold=10) -> str:
+    """Trim input sequence according to suffix and prefix number of indels
+
+    Args:
+        prefix_type (str): type of edits at the beginning of the cigar, ("I" or "D")
+        prefix_len (int): length of prefix edits
+        suffix_type (str):type of edits at the end of the cigar, ("I" or "D")
+        suffix_len (int): length of suffix edits
+        seq (str): input sequence to be trimmed
+        num (int): ranking of the sequence. If a sequence is the first in alignment, deletion means that removing. 
+        Otherwise, insertion means trimming.
+        threshold (int): maximum number of bases trimmed
+
+    Returns:
+        str: trimmed sequence
+    """
+    assert num == 1 or num == 2, "trim_seq(): unknown ranking of input sequence!"
+
+    if num == 2:
+        if prefix_type == "D":
+            seq = seq[min(threshold, prefix_len):]
+        if suffix_type == "D":
+            seq = seq[:-min(threshold, suffix_len)]
+    if num == 1:
+        if prefix_type == "I":
+            seq = seq[min(threshold, prefix_len):]
+        if suffix_type == "I":
+            seq = seq[:-min(threshold, suffix_len)]
+    return seq
+
+def trim_from_cigar(cigar_str, sequence1, sequence2, threshold=10, mode="lenient"):
+    """Given two sequences and the cigar string describing the alignment between them, trim the start and end, if the 
+    alignment starts and/or ends with indels.
+
+    Args:
+        cigar_str (str): cigar_string
+        sequence1(str): aligned sequence 1 to be trimmed
+        sequence2(str): aligned sequence 2 to be trimmed
+        threshold(int): maximum number of bases trimmed
+        mode(str): strict or lenient
+            - strict: only trim if the trimmed result will be all matches
+            - lenient: trim whenever there are indels at the edge of the strings
+    Returns:
+        str: trimmed sequence 1
+        str: trimmed sequence 2
+    """
+
+    (prefix_type, prefix_length), (suffix_type, suffix_length) = get_prefix_and_suffix_indels(cigar_str, mode)
+
+    if prefix_length == 0 and suffix_length == 0:
+        return sequence1, sequence2
+
+    global total_trim
```

Comment:
I worry global variables in python are almost always evil :)

It feels like a language flaw of Python

Wouldn't these be more clear in the main() function perhaps?

Reply:
I can make changes after things are finalized. For now it is the best and most convenient way to keep track of numbers for debugging. If these variables are declared in the main function, they will not be accessible by the other functions unless they are passed as arguments, which makes less sense since they are merely counters. In the final version I don't think we need to report these numbers so we can just remove these.